php/class-customize-snapshot-manager.php

```diff
@@ -0,0 +1,559 @@
+		// Don't show for users who can't access the customizer or when in the admin.
+		if ( ! current_user_can( 'customize' ) || is_admin() ) {
+			return;
+		}
+
+		$current_url = $this->current_url();
+		parse_str( parse_url( $current_url, PHP_URL_QUERY ), $query_vars );
+
+		$args = array();
+		$uuid = isset( $query_vars['customize_snapshot_uuid'] ) ? $query_vars['customize_snapshot_uuid'] : null;
+		$scope = isset( $query_vars['scope'] ) ? $query_vars['scope'] : 'dirty';
+
+		if ( $uuid && $this->snapshot->is_valid_uuid( $uuid ) ) {
+			$args['customize_snapshot_uuid'] = $uuid;
+			$args['scope'] = $scope;
+		}
+
+		$args['url'] = urlencode( $current_url );
+		$customize_url = add_query_arg( $args, wp_customize_url() );
+
+		$wp_admin_bar->add_menu(
+			array(
+				'id'     => 'customize',
+				'title'  => __( 'Customize' ),
+				'href'   => $customize_url,
+				'meta'   => array(
+					'class' => 'hide-if-no-customize',
+				),
+			)
+		);
+		add_action( 'wp_before_admin_bar_render', 'wp_customize_support_script' );
+	}
+
+	/**
+	 * Underscore (JS) templates for dialog windows.
+	 */
+	public function render_templates() {
+		?>
+		<script type="text/html" id="tmpl-snapshot-button">
+			<button id="snapshot-button" class="dashicons dashicons-share">
+				<span class="screen-reader-text">{{ data.buttonText }}</span>
+			</button>
+		</script>
+
+		<script type="text/html" id="tmpl-snapshot-dialog-share-link">
+			<div id="snapshot-dialog-share-link" title="{{ data.title }}">
+				<a href="{{ data.url }}" target="_blank">{{ data.url }}</a>
+			</div>
+		</script>
+
+		<script type="text/html" id="tmpl-snapshot-dialog-share-error">
+			<div id="snapshot-dialog-share-error" title="{{ data.title }}">
+				<p>{{ data.message }}</p>
+			</div>
+		</script>
+
+		<script type="text/html" id="tmpl-snapshot-dialog-form">
+			<div id="snapshot-dialog-form" title="{{ data.title }}">
+				<form>
+					<fieldset>
+						<# if ( data.is_preview ) { #>
+							<p>{{ data.message }}</p>
+							<input type="hidden" value="{{ data.scope }}" name="scope">
+						<# } else { #>
+							<label for="type-0">
+								<input id="type-0" type="radio" checked="checked" value="dirty" name="scope">{{ data.dirtyLabel }}
+							</label>
+							<br>
+							<label for="type-1">
+								<input id="type-1" type="radio" value="full" name="scope">{{ data.fullLabel }}
+							</label>
+							<br>
+						<# } #>
+						<input type="submit" tabindex="-1" style="position:absolute; top:-5000px" />
+					</fieldset>
+				</form>
+			</div>
+		</script>
+		<?php
+	}
+
+	/**
+	 * Check if the setting can be previewed.
+	 *
+	 * @param \WP_Customize_Setting $setting A WP_Customize_Setting derived object.
+	 * @param array                 $values  All settings' values in the snapshot.
+	 * @return bool
+	 */
+	public function can_preview( $setting, $values ) {
+		if ( ! ( $setting instanceof \WP_Customize_Setting ) && ! is_subclass_of( $setting, 'WP_Customize_Setting' ) ) {
```

Comment:
@valendesigns using `is_subclass_of` is redundant here, as `instanceof` also works for parent classes:

> instanceof can also be used to determine whether a variable is an instantiated object of a class that inherits from a parent class:
http://php.net/manual/en/language.operators.type.php